pandas/core/frame.py

```diff
@@ -6618,16 +6619,39 @@ def _gotitem(
         **_shared_doc_kwargs,
     )
     @Appender(_shared_docs["aggregate"])
-    def aggregate(self, func, axis=0, *args, **kwargs):
+    def aggregate(self, func=None, axis=0, *args, **kwargs):
         axis = self._get_axis_number(axis)
 
+        relabeling, func, columns, order = reconstruct_func(func, *args, **kwargs)
+        if relabeling:
+            kwargs = {}
+
         result = None
         try:
             result, how = self._aggregate(func, axis=axis, *args, **kwargs)
         except TypeError:
             pass
         if result is None:
             return self.apply(func, axis=axis, args=args, **kwargs)
+
+        if relabeling:
+            reordered_indexes = [
+                pair[0] for pair in sorted(zip(columns, order), key=lambda t: t[1])
+            ]
+
+            # when there are more than one column being used in aggregate, the order
+            # of result will be reversed, and in case the func is not used by other
+            # columns, there might be NaN values, so separate these two cases
+            reordered_result = DataFrame(index=columns)
+            idx = 0
+            for col, funcs in func.items():
```

Comment:
this is extremely inefficient.

instead try appending to a list and concat at the end.

why do you need .values? that is not very idiomatic here, nor is the dropna

Reply:
you can also just reorder the indicies and do an indexing selection would be way better

Reply:
yeah, I tried to reduce half of the code in this loop, i am pretty sure the previous version was due to some order issue.
However, this seems still a bit complex, but this is to keep two things unchanged: 1. the order to first occurrence of columns to be aggregated, 2. the order of the new named column occurrence.

for instance in the example below, i think a correct behaviour should be:
```python
df = pd.DataFrame({"A": [1, 2, 1, 2], "B": [1, 2, 3, 4], "C": [3,4,5,6]})
df.agg(ab=("C", np.max), foo=("A", max), bar=("B", "mean"), cat=("A", "min"), 
            dat=("B", "max"), f=("A", "max"), g=("C", "min"))
```
the column of new aggregated df is : `['C', 'A', 'B']` other than `['A', 'B', 'C']` since it is the order of the first occurrence of them. And the index should be `['ab', 'foo', 'bar', 'cat', 'dat', 'f', 'g']` since it is the order in the agg function.

Any advice to simply the code would be much appreciated!